### DISCUSSION TITLE
How to use POST /optimize?

### DISCUSSION THREAD

**Discussion by PirminHndlg** [Akkudoktor-EOS/EOS#301] [Q&A]
Hello,

I am new here and wanted to try out this amazing project! I cloned the project, installed everything and played around with the swegger site. Unfortunately I could not find any information about the meaning of the json values. I was trying to get an optimisation with this json data:

```curl -X 'POST' \
  'http://192.168.178.193:8503/optimize?start_hour=2' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "ems": {
    "pv_prognose_wh": [
      350
    ],
    "strompreis_euro_pro_wh": [
      0.0003
    ],
    "einspeiseverguetung_euro_pro_wh": [
      0.00007
    ],
    "preis_euro_pro_wh_akku": 0.0003,
    "gesamtlast": [
      240
    ]
  },
  "pv_akku": {
    "kapazitaet_wh": 10000,
    "lade_effizienz": 0.88,
    "entlade_effizienz": 0.88,
    "max_ladeleistung_w": 5000,
    "start_soc_prozent": 0,
    "min_soc_prozent": 0,
    "max_soc_prozent": 100
  },
  "wechselrichter": {
    "max_leistung_wh": 800
  },
  "eauto": {
    "kapazitaet_wh": 80000,
    "lade_effizienz": 0.88,
    "entlade_effizienz": 1,
    "max_ladeleistung_w": 1,
    "start_soc_prozent": 0,
    "min_soc_prozent": 0,
    "max_soc_prozent": 100
  },
  "dishwasher": {
    "consumption_wh": 500,
    "duration_h": 3
  },
  "temperature_forecast": [
    0
  ],
  "start_solution": [
    1,
    2
  ]
}'
```

But I got a 500 error:
```
eos-1  | Time evaluate inner: 13.4691 sec.
eos-1  | INFO:     192.168.178.204:49285 - "POST /optimize?start_hour=2 HTTP/1.1" 500 Internal Server Error
eos-1  | ERROR:    Exception in ASGI application
eos-1  | Traceback (most recent call last):
eos-1  |   File "/usr/local/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 409, in run_asgi
...
```
Did I do something wrong in the json? I am not sure what 'start_solution' means, but it needs an array of length 2.

Thank you for your responses.

**Comment by Lasall** [comment]
There is currently missing input validation (thanks for finding this bug). Try to add 48 (=prediction_hours) values for gesamtlast, pv_prognose_wh, strompreis_euro_pro_wh (and einspeiseverguetung_euro_pro_wh or just use a float there instead of list, and also temperature_forecast or remove that). Currently it throws the error when the len(values) < start_hour where start_hour is the current hour of the day, e.g. at 23:44 -> 23

Take a look at tests/testdata/optimize_input*.json for example data.
